python/ray/_private/services.py

```diff
@@ -919,7 +919,23 @@ def start_redis(node_ip_address,
         # If no port is given, fallback to default Redis port for the primary
         # shard.
         if port is None:
-            port = ray_constants.DEFAULT_PORT
+            default_port = ray_constants.DEFAULT_PORT
+            if sys.platform == "win32":
+                port = default_port
+                num_choices = 200
+                a_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+                while num_choices > 0:
+                    location = (node_ip_address, port)
+                    try:
+                        a_socket.bind(location)
+                        break
+                    except socket.error as e:
+                        if e.errno == errno.EADDRINUSE:
+                            port += 1
+                            num_choices -= 1
+                a_socket.close()
+            else:
+                port = default_port
```

Comment:
This first starts from `6379` and checks if it available by creating a temporary socket. This socket tries to bind to the (`node_ip_address`, `port`) and if it's successful then the loop breaks and the port is selected. Otherwise it moves on to the next port ( + 1) and keeps on trying till `num_choices` times. In the end the temporary socket is closed. 

Let me know if this sounds fine. @wuisawesome @pcmoritz

Reply:
The above approach will fail in one case. When two subprocess are started simultaneously (see example below). In that case, both subprocesses will be able to see `6379` as an available port and hence both will be using the same port despite the fact that the cause of the issue is usage of same port by multiple instances. So there are two ways in my mind to deal with this,

1. Select initial port as `6379` on a chance basis. For example, `port = default_port if random.random() >= 0.5 else default_port + 1`.
2. Or break from the `while num_choices > 0` loop on a chance basis so that only one of the two simultaneously running instances is able to grab `6379` as the first port.

Reply:
> The above approach will fail in one case.

Agreed, let's add a TODO, but that seems fine.

Reply:
Okay. It's really tricky case. Simplest way is to not start from default_port but that doesn't preserve the cross platform behaviour. I will add a TODO anyways.

Reply:
A work around is to add, `time.sleep(random.uniform(0, 1))` before `a_socket.bind(location)` so that only one process/instance is able to grab `6379` because, two processes will wake up at different times due to the random sleep duration and hence, one will be quicker than the other. Afterwards, we pick next port randomly so there will be no collisions (or at least the chances will be extremely low and even if that happens we have the random sleep time to come out from that). 

Other way is to add a note in the documentation in ray for Windows users. They can start their different ray instances at some gaps instead of calling all the processes simultaneously.

Reply:
The sad part is that this issue is only in Windows. On macOS, only one socket is able to see a common port as free, the other one throws already in use error.